OpenRA.Mods.Common/Traits/Upgrades/UpgradableTrait.cs

```diff
@@ -56,33 +64,63 @@ public virtual void RulesetLoaded(Ruleset rules, ActorInfo ai)
 	public abstract class UpgradableTrait<InfoType> : IUpgradable, IDisabledTrait, ISync where InfoType : UpgradableTraitInfo
 	{
 		public readonly InfoType Info;
-		public IEnumerable<string> UpgradeTypes { get { return Info.UpgradeTypes; } }
+		readonly Dictionary<string, bool> conditions = new Dictionary<string, bool>();
+
+		IEnumerable<string> IUpgradable.UpgradeTypes
+		{
+			get
+			{
+				if (Info.EnableCondition != null)
+					return Info.EnableCondition.Variables;
+
+				return Info.UpgradeTypes;
+			}
+		}
+
 		[Sync] public bool IsTraitDisabled { get; private set; }
 
 		public UpgradableTrait(InfoType info)
 		{
 			Info = info;
-			IsTraitDisabled = info.UpgradeTypes != null && info.UpgradeTypes.Count > 0 && info.UpgradeMinEnabledLevel > 0;
+
+			// TODO: Set initial state from a ConditionsInit once that exists
+			if (info.EnableCondition != null)
+				IsTraitDisabled = !info.EnableCondition.Evaluate(conditions);
+			else
+				IsTraitDisabled = info.UpgradeTypes != null && info.UpgradeTypes.Count > 0 && info.UpgradeMinEnabledLevel > 0;
 		}
 
-		public bool AcceptsUpgradeLevel(Actor self, string type, int level)
+		bool IUpgradable.AcceptsUpgradeLevel(Actor self, string type, int level)
 		{
+			if (Info.EnableCondition != null)
+				return level == 1;
+
 			return level > 0 && level <= Info.UpgradeMaxAcceptedLevel;
 		}
 
-		public void UpgradeLevelChanged(Actor self, string type, int oldLevel, int newLevel)
+		void IUpgradable.UpgradeLevelChanged(Actor self, string type, int oldLevel, int newLevel)
 		{
-			if (!Info.UpgradeTypes.Contains(type))
-				return;
+			var wasDisabled = IsTraitDisabled;
+
+			if (Info.EnableCondition != null)
+			{
+				conditions[type] = newLevel > 0;
+				IsTraitDisabled = !Info.EnableCondition.Evaluate(conditions);
+			}
+			else
+			{
+				if (!Info.UpgradeTypes.Contains(type))
+					return;
 
-			// Restrict the levels to the allowed range
-			oldLevel = oldLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
-			newLevel = newLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
-			if (oldLevel == newLevel)
-				return;
+				// Restrict the levels to the allowed range
+				oldLevel = oldLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
+				newLevel = newLevel.Clamp(0, Info.UpgradeMaxAcceptedLevel);
+				if (oldLevel == newLevel)
+					return;
```

Comment:
This til the end of the block could be:

```csharp
if (oldLevel != newLevel)
    IsTraitDisabled = newLevel < Info.UpgradeMinEnabledLevel || newLevel > Info.UpgradeMaxEnabledLevel;
```

This doesn't matter in the grand scheme of course.